pkg/container/kubernetes_runner.go

```diff
@@ -569,3 +568,91 @@ func (k *k8sLoader) LoadImage(ctx context.Context, layer ggcrv1.Layer, arch apko
 
 	return ref.String(), nil
 }
+
+type kubernetesRunnerTarFetcher struct {
+	reader     *io.PipeReader
+	outStream  *io.PipeWriter
+	bytesRead  uint64
+	retries    int
+	maxRetries int
+
+	podName      string
+	podNamespace string
+
+	restConfig *rest.Config
+	client     kubernetes.Interface
+	logger     log.Logger
+}
+
+// newKubernetesRunnerTarFetcher creates and "opens" the tar pipe to the remote pod, the pod must be running
+func newKubernetesRunnerTarFetcher(cfg *rest.Config, client kubernetes.Interface, logger log.Logger, namespace string, name string) *kubernetesRunnerTarFetcher {
+	t := &kubernetesRunnerTarFetcher{
+		maxRetries:   10,
+		podName:      name,
+		podNamespace: namespace,
+		restConfig:   cfg,
+		client:       client,
+		logger:       logger,
+	}
+	t.initReadFrom(0)
+	return t
+}
+
+func (t *kubernetesRunnerTarFetcher) initReadFrom(n uint64) {
+	t.reader, t.outStream = io.Pipe()
+
+	execfn := func() error {
+		req := t.client.
+			CoreV1().
+			RESTClient().
+			Post().
+			Resource("pods").
+			Name(t.podName).
+			Namespace(t.podNamespace).
+			SubResource("exec").
+			VersionedParams(&corev1.PodExecOptions{
+				Container: kubernetesBuilderPodWorkspaceContainerName,
+				Command:   []string{"/bin/sh", "-c", fmt.Sprintf("tar -cf - -C %s %s | tail -c+%d", runnerWorkdir, "melange-out", n)},
+				Stdout:    true,
+				Stderr:    true,
+			}, scheme.ParameterCodec)
+
+		exec, err := remotecommand.NewSPDYExecutor(t.restConfig, "POST", req.URL())
+		if err != nil {
+			return fmt.Errorf("failed to create remote command executor: %v", err)
+		}
+
+		return exec.StreamWithContext(context.Background(), remotecommand.StreamOptions{
```

Comment:
Please propagate context 🙏

Reply:
I missed this thanks! this also made me realize I missed passing the actual retry 🤯 

will refactor things slightly to make sure context is passed while still implementing `io.Reader`